source/Engine.cpp

```diff
@@ -100,35 +100,16 @@ namespace {
 		return Radar::UNFRIENDLY;
 	}
 
-	template <class Type>
-	void Prune(vector<Type> &objects)
+	template <template<class, class> class Container, class Alloc, class Inner>
+	void Prune(Container<shared_ptr<Inner>, Alloc> &objects)
 	{
-		// First, erase any of the old objects that should be removed.
-		typename vector<Type>::iterator in = objects.begin();
-		while(in != objects.end() && !in->ShouldBeRemoved())
-			++in;
-
-		typename vector<Type>::iterator out = in;
-		while(in != objects.end())
-		{
-			if(!in->ShouldBeRemoved())
-				*out++ = std::move(*in);
-			++in;
-		}
-		if(out != objects.end())
-			objects.erase(out, objects.end());
+		erase_if(objects, [](const auto &ptr){return ptr->ShouldBeRemoved();});
 	}
 
-	template <class Type>
-	void Prune(list<shared_ptr<Type>> &objects)
+	template <template<class, class> class Container, class Type, class Alloc>
+	void Prune(Container<Type, Alloc> &objects)
 	{
-		for(auto it = objects.begin(); it != objects.end(); )
-		{
-			if((*it)->ShouldBeRemoved())
-				it = objects.erase(it);
-			else
-				++it;
-		}
+		erase_if(objects, [](const auto &object){return object.ShouldBeRemoved();});
 	}
```

Comment:
Instead of using all these templates we can just do
```cpp
	constexpr auto PrunePtrs = [](auto &objects) { erase_if(objects,
	   [](const auto &obj) { return obj->ShouldBeRemoved(); }); };
	constexpr auto Prune = [](auto &objects) { erase_if(objects,
	   [](const auto &obj) { return obj.ShouldBeRemoved(); }); };
```
Before the pruning steps, and use these lambdas instead.